lib/ui/painting/codec.cc

```diff
@@ -397,49 +436,58 @@ int MultiFrameCodec::repetitionCount() {
   return repetitionCount_;
 }
 
-sk_sp<SkImage> MultiFrameCodec::GetNextFrameImage(
-    fml::WeakPtr<GrContext> resourceContext) {
-  // Populate this bitmap from the cache if it exists
-  DecodedFrame& cacheEntry = *frameBitmaps_[nextFrameIndex_];
-  SkBitmap bitmap =
-      cacheEntry.bitmap_ != nullptr ? *cacheEntry.bitmap_ : SkBitmap();
-  if (!bitmap.getPixels()) {  // We haven't decoded this frame yet
-    const SkImageInfo info = codec_->getInfo().makeColorType(kN32_SkColorType);
-    bitmap.allocPixels(info);
-
-    SkCodec::Options options;
-    options.fFrameIndex = nextFrameIndex_;
-    const int requiredFrame = frameInfos_[nextFrameIndex_].fRequiredFrame;
-    if (requiredFrame != SkCodec::kNoFrame) {
-      const SkBitmap* requiredBitmap =
-          frameBitmaps_[requiredFrame]->bitmap_.get();
-      if (requiredBitmap == nullptr) {
-        FML_LOG(ERROR) << "Frame " << nextFrameIndex_ << " depends on frame "
-                       << requiredFrame << " which has not been cached.";
-        return NULL;
-      }
-
-      if (requiredBitmap->getPixels() &&
-          copy_to(&bitmap, requiredBitmap->colorType(), *requiredBitmap)) {
-        options.fPriorFrame = requiredFrame;
-      }
+bool MultiFrameCodec::DecodePixels(SkBitmap* bitmap, int frame) {
+  if (bitmap->getPixels() != nullptr) {
+    // Don't decode this twice.
+    return true;
+  }
+
+  const SkImageInfo info = codec_->getInfo().makeColorType(kN32_SkColorType);
+  bitmap->allocPixels(info);
+
+  SkCodec::Options options;
+  options.fFrameIndex = frame;
+  const int requiredFrame = frameInfos_[frame].fRequiredFrame;
+  if (requiredFrame != SkCodec::kNoFrame) {
+    DecodedFrame& requiredCacheEntry = *frameBitmaps_[requiredFrame];
+    SkBitmap requiredBitmap;
+    if (!requiredCacheEntry.CachedBitmapExists()) {
```

Comment:
Shouldn't the required frame always be cached at this point? (we are always caching required frames, and only evicting them after their last usage AFAIU)